Fix training IR bug by changing passes order

tugsbayasgalan opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #138517
* #138516
* #138261
* __->__ #138292
* #138266

Inserting runtime_assertions cause gm to have different names but the graph signature was populated earlier. To avoid this kind of errors in the future, I refactored these steps into a helper function. 

Differential Revision: [D64576251](https://our.internmc.facebook.com/intern/diff/D64576251)

tugsbayasgalan committed:
Fix training IR bug by changing passes order

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D64576251).

avikchaudhuri commented on torch/_export/utils.py:
```diff
@@ -529,6 +533,34 @@ def nodes_filter(nodes: List[torch.fx.Node], node_call_back) -> List[torch.fx.No
+    # update output specs
+    gm.recompile()
+    graph_signature.user_outputs = _graph_output_names(gm)
+    return gm, graph_signature
```
why return gm, graph_signature if you're not using it below?

tugsbayasgalan replied:
Yeah it can return None. But maybe i should return and replace to make it more clear

tugsbayasgalan committed:
Update on "Fix training IR bug by changing passes order"


Inserting runtime_assertions cause gm to have different names but the graph signature was populated earlier. To avoid this kind of errors in the future, I refactored these steps into a helper function. 

Differential Revision: [D64576251](https://our.internmc.facebook.com/intern/diff/D64576251)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D64576251).

tugsbayasgalan committed:
Update on "Fix training IR bug by changing passes order"


Inserting runtime_assertions cause gm to have different names but the graph signature was populated earlier. To avoid this kind of errors in the future, I refactored these steps into a helper function. 

Differential Revision: [D64576251](https://our.internmc.facebook.com/intern/diff/D64576251)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D64576251).

tugsbayasgalan committed:
Update on "Fix training IR bug by changing passes order"


Inserting runtime_assertions cause gm to have different names but the graph signature was populated earlier. To avoid this kind of errors in the future, I refactored these steps into a helper function. 

Differential Revision: [D64576251](https://our.internmc.facebook.com/intern/diff/D64576251)

[ghstack-poisoned]

tugsbayasgalan commented:
@pytorchbot rebase

tugsbayasgalan committed:
Update

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Fix training IR bug by changing passes order"


Inserting runtime_assertions cause gm to have different names but the graph signature was populated earlier. To avoid this kind of errors in the future, I refactored these steps into a helper function. 

Differential Revision: [D64576251](https://our.internmc.facebook.com/intern/diff/D64576251)

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Fix training IR bug by changing passes order"


Inserting runtime_assertions cause gm to have different names but the graph signature was populated earlier. To avoid this kind of errors in the future, I refactored these steps into a helper function. 

Differential Revision: [D64576251](https://our.internmc.facebook.com/intern/diff/D64576251)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D64576251).

tugsbayasgalan commented:
@pytorchbot merge -f "Landed internally"
